Blogbeitrag: Open Platform Communication (#284)
adessoSE/devblog: _posts/2023-09-08-open-platform-communication.md

author_ids: [cjohn]                   # Pflichtfeld. Es muss in der "authors.yml" einen Eintrag mit diesen Namen geben.
categories: [Branchen]                # Pflichtfeld. Maximal eine der angegebenen Kategorien verwenden.
tags: [Open Platform Communication, OPC, Applications, Hardware, Prozesssteuerung] # Bitte auf Großschreibung achten.
---

Maschinen sprechen miteinander.
Sie stimmen Prozessschritte ab, nehmen Anweisungen entgegen und melden ihren Zustand.
Eine ihrer Sprachen ist Open Platform Communication.

# Wir müssen reden

Industriegeräte tauschen ständig Information aus.
Sie fragen nach Sensorwerten oder dem Arbeitsfortschritt.
Sie erteilen einander Aufträge und sie stoppen, sobald anderswo im Prozess ein Fehler auftritt.

Als Beispiel soll hier ein Windpark mit zwei Turbinen herhalten.
Mit ihren Sensoren überwachen die Windräder sich selbst, doch allein reparieren können sie sich natürlich nicht.

![Fallbeispiel](/assets/images/posts/open-platform-communication/windrad.png)

Deshalb müssen bevorstehender Wartungsbedarf und akute Schäden sofort an die Technikabteilung gemeldet werden.
Letztere antwortet darauf mit Steuerbefehlen, die der Windpark umsetzen muss.

> Für den stabilen Netzbetrieb ist es außerdem wichtig, die aktuelle Einspeiseleistung genau zu kennen und gegebenenfalls herunter regeln zu können.

Reviewer: ```suggestion
Für den stabilen Netzbetrieb ist es außerdem wichtig, die aktuelle Einspeiseleistung genau zu kennen und gegebenenfalls herunterregeln zu können.
```

## Alarms & Conditions

Um Probleme sofort mitzubekommen, wird ein Publish-Subscribe-Verfahren benötigt.
Dabei meldet sich der Client einmal für Ereignisse eines Knotens an, von da an wird er über jede Zustandsänderung aktiv informiert.

In der Abbildung ist "Alarm" ein solcher Ereignisknoten. Er enthält von jeder Turbine einen Notifier.
Dieselben Notifier "T1-Error" und "T2-Error" finden sich auch unterhalb der jeweiligen Turbine.
Sie haben jeweils zwei Elternknoten - Turbine und Windpark - damit Clients wahlweise nur eine Turbine oder den ganzen Park beobachten können.
Beide Alarmereignisse sind normalerweise inaktiv. Nur wenn ein Fehlercode vorliegt, wird der Alarm aktiv.

So löst AE den zweiten Anwendungsfall, alle interessierten Clients werden im Problemfall sofort gewarnt.
Die Unterprotokolle DA und AE sind also ausreichend, um einen Windpark fernzusteuern.

# Gründe für eine Migration von OPC Classic zu UA

Obwohl OPC UA seit mehr als 15 Jahren existiert, bauen viele produktive Systeme nach wie vor auf OPC Classic auf.
Denn um die Software umzustellen, müssen alle beteiligten Geräte und Anwendungen OPC UA unterstützen.

> Bei der Steuersoftware für Maschinen, die über Jahrzehnte in der Produktion oder in Kraftwerken genutzt werden, ist kaum mit tiefgreifenden Updates zu rechnen.

Reviewer: ```suggestion
Bei der Steuersoftware für Maschinen, die über Jahrzehnte in der Produktion oder in Kraftwerken genutzt wird, ist kaum mit tiefgreifenden Updates zu rechnen.
```

> Andere Geräte sind systemkritisch, so dass sie nicht für ein riskantes Upgrade mehrerer Software-Komponenten abgeschaltet werden sollen.

Reviewer: ```suggestion
Andere Geräte sind systemkritisch, sodass sie nicht für ein riskantes Upgrade mehrerer Softwarekomponenten abgeschaltet werden sollen.
```

Ältere Mitlesende erinnern sich noch, wie jedes eigene Programm eine Word- oder Excel-Schnittstelle anfordern konnte.
Das fernsteuerbare Programm - zu Beispiel Excel - registrierte sich genau einmal als COM-Komponente in einer zentralen Registry.
Client-Programme schlugen dort nach und forderten vom Betriebssystem einen COM-Server vom Typ Excel an.
Im Hintergrund sorgten Remote Procedure Calls (RPC) dafür, dass der Client das Serverprogramm steuern konnte, fast als liefen beide im selben Prozess.

Mit Distributed COM (DCOM) kam später Ortstransparenz dazu, so dass die COM-Registry auch auf entfernte Rechner verweisen konnte.
Der Client sollte DCOM-Komponenten genauso verwenden, als wären sie Programmteile im eigenen Prozess - im Idealfall, ohne mitzubekommen, auf welchem Rechner sie tatsächlich liefen.
In der Praxis führte das oft zu komplizierten Konfigurationen.
Nach und nach wurden auch gravierende Sicherheitslücken bekannt.
Schließlich riet Microsoft von der Verwendung ab und präsentierte als Nachfolger die Windows Communication Foundation (WCF).
Seit dem Release von .NET 5 ist selbst diese wieder abgekündigt.

> Dazu kommt, dass Windows in hardware-nahen Systemen an Relevanz verloren hat.

Reviewer: ```suggestion
Dazu kommt, dass Windows in hardwarenahen Systemen an Relevanz verloren hat.
```
